external/sql/storm-sql-external/storm-sql-hdfs/src/test/org/apache/storm/sql/hdfs/TestHdfsDataSourcesProvider.java

```diff
@@ -0,0 +1,152 @@
+import java.io.IOException;
+import java.net.URI;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Properties;
+
+import static org.apache.storm.hdfs.trident.HdfsState.HdfsFileOptions;
+import static org.mockito.Matchers.argThat;
+import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+
+public class TestHdfsDataSourcesProvider {
+  private static final List<FieldInfo> FIELDS = ImmutableList.of(
+      new FieldInfo("ID", int.class, true),
+      new FieldInfo("val", String.class, false));
+  private static final List<String> FIELD_NAMES = ImmutableList.of("ID", "val");
+  private static final JsonSerializer SERIALIZER = new JsonSerializer(FIELD_NAMES);
+  private static final Properties TBL_PROPERTIES = new Properties();
+
+  private static String hdfsURI;
+  private static MiniDFSCluster hdfsCluster;
+
+  static {
+    TBL_PROPERTIES.put("hdfs.file.path", "/unittest");
+    TBL_PROPERTIES.put("hdfs.file.name", "test1.txt");
+    TBL_PROPERTIES.put("hdfs.rotation.time.seconds", "120");
+  }
+
+  @Before
+  public void setup() throws Exception {
+    Configuration conf = new Configuration();
+    conf.set("fs.trash.interval", "10");
+    conf.setBoolean("dfs.permissions", true);
+    File baseDir = new File("./target/hdfs/").getAbsoluteFile();
+    FileUtil.fullyDelete(baseDir);
+    conf.set(MiniDFSCluster.HDFS_MINIDFS_BASEDIR, baseDir.getAbsolutePath());
+
+    MiniDFSCluster.Builder builder = new MiniDFSCluster.Builder(conf);
+    hdfsCluster = builder.build();
+    hdfsURI = "hdfs://localhost:" + hdfsCluster.getNameNodePort() + "/";
+  }
+
+  @After
+  public void shutDown() throws IOException {
+    hdfsCluster.shutdown();
+  }
+
+  @SuppressWarnings("unchecked")
+  @Test
+  public void testHdfsSink() {
+    ISqlTridentDataSource ds = DataSourcesRegistry.constructTridentDataSource(
+            URI.create(hdfsURI), null, null, TBL_PROPERTIES, FIELDS);
+    Assert.assertNotNull(ds);
+
+    ISqlTridentDataSource.SqlTridentConsumer consumer = ds.getConsumer();
+
+    Assert.assertEquals(HdfsStateFactory.class, consumer.getStateFactory().getClass());
+    Assert.assertEquals(HdfsUpdater.class, consumer.getStateUpdater().getClass());
+
+    HdfsState state = (HdfsState) consumer.getStateFactory().makeState(Collections.emptyMap(), null, 0, 1);
+    StateUpdater stateUpdater = consumer.getStateUpdater();
+
+    HdfsFileOptions options = mock(HdfsFileOptions.class);
+    FSDataOutputStream out = mock(FSDataOutputStream.class);
+    Whitebox.setInternalState(state, "options", options);
+    Whitebox.setInternalState(options, "out", out);
+
+    List<TridentTuple> tupleList = mockTupleList();
+
+    for (TridentTuple t : tupleList) {
+      stateUpdater.updateState(state, Collections.singletonList(t), null);
+      try {
+        verify(out).write(argThat(new HdfsArgMatcher(t)));
```

Comment:
@vesense 
testHdfsSink() is failing consistency. I guess options are mocked so options.execute() doesn't work as expected. You need to set options.execute() to run real method, and provide other needed fields as well.